File: code/210_solution.cpp
class Solution {
public:
    vector<int> findOrder(int N, vector<vector<int>>& prerequisites) {
        bool taken[100000]={false}, path[100000]={false};  // O(N)
        // graph processing  edge list -> adj list
        vector<vector<int>> next_courses(N);
        for (auto t:prerequisites) next_courses[t[1]].push_back(t[0]); // O(E), O(E)
        vector<int> order;
        // dfs function return cycle or not
        function<bool(int)> dfs = [&](int c) {
            if (taken[c]) return false;
            if (path[c]) return true;
            path[c] = true;
            
            for (int next:next_courses[c]) {
                bool has_cycle = dfs(next);
                if (has_cycle) return true;
            }
            taken[c] = true;
            path[c] = false;
            order.push_back(c);
            return false;
        };
        
        // try dfs search to detect cycle
        for (int c=0;c<N;c++) {  // O(N) time
            if (taken[c]) continue;
            if (dfs(c)) return {};  // O(E) time
        };
        
        reverse(order.begin(), order.end());
        return order;        
    }
};

File: code/210_solution.py
class Solution:
    def findOrder(self, N: int, prerequisites: List[List[int]]) -> List[int]:
        # graph processing
        out_edges = [[] for _ in range(N)]
        in_degrees = [0 for _ in range(N)]
        for course, prereq in  prerequisites:
            out_edges[prereq].append(course)
            in_degrees[course] += 1
        
        # bfs search for feasibilty as well as order
        queue = [course for course, prereq in enumerate(in_degrees) if not prereq]
        queue = deque(queue)
        taken = []
        while queue:
            course = queue.popleft()
            taken.append(course)
            for next_course in out_edges[course]:
                in_degrees[next_course] -= 1
                if in_degrees[next_course] == 0: queue.append(next_course)
        return taken if len(taken) == N else []